portindex2json/portindex2json.tcl

```diff
@@ -12,6 +12,61 @@
 
 package require json::write
 
+proc tcl2json_list {tcl_list} {
+    set new_tcl_list {}
+    foreach element $tcl_list {
+        lappend new_tcl_list [::json::write string $element]
+    }
+    return [::json::write array {*}$new_tcl_list]
+}
+
+proc get_maintainer_array {maintainer} {
+    set detail_list [split $maintainer @]
+    if {[string first : [lindex $detail_list 0]] != -1} {
+        set email_list [split [lindex $detail_list 0] :]
+        set name [string trimright [lindex $email_list 1]]
+        set email(name) [::json::write string $name]
+        set email(domain) [::json::write string [lindex $email_list 0]]
+        set email_subobject [::json::write object {*}[array get email]]
+    } else {
+        set name [string trimright [lindex $detail_list 0]]
+        set email(name) [::json::write string $name]
+        set email(domain) [::json::write string "macports.org"]
+        set email_subobject [::json::write object {*}[array get email]]
```

Comment:
`email(name)` and `email_subobject` seem to be the same, so they could probably go out of the if-else loop?

I still want to test this (it would probably be nice to have some unit tests) to make sure that different scenarios work ok (if there is no email, if there is no github handle, if github comes before email, if it comes after email etc.)

Reply:
> `email(name)` and `email_subobject` seem to be the same, so they could probably go out of the if-else loop?

They are not same, `email(name)` is used to set the key `name`. And using different keys, the object `email_subobject` is compiled. 
Am I understanding your message correctly?

> I still want to test this (it would probably be nice to have some unit tests) to make sure that different scenarios work ok (if there is no email, if there is no github handle, if github comes before email, if it comes after email etc.)

I am afraid if won't work if the formats `{name:domain @github}` or `{name @github}` are altered. Because it totally relies on splitting the string at `@`.

Reply:
What I meant was the following:
```diff
 if {[string first : [lindex $detail_list 0]] != -1} {
     set email_list [split [lindex $detail_list 0] :]
     set name [string trimright [lindex $email_list 1]]
-    set email(name) [::json::write string $name]
     set email(domain) [::json::write string [lindex $email_list 0]]
-    set email_subobject [::json::write object {*}[array get email]]
 } else {
     set name [string trimright [lindex $detail_list 0]]
-    set email(name) [::json::write string $name]
     set email(domain) [::json::write string "macports.org"]
-    set email_subobject [::json::write object {*}[array get email]]
-}
+set email(name) [::json::write string $name]
+set email_subobject [::json::write object {*}[array get email]]
```
I didn't test so maybe some variables need to be defined outside of the loop in order not to loose the scope.

My question is: does the code still work if you change `{domain:name @github}` into `{@github domain:name}`?

Reply:
> What I meant was the following:
> 
> ```diff
>  if {[string first : [lindex $detail_list 0]] != -1} {
>      set email_list [split [lindex $detail_list 0] :]
>      set name [string trimright [lindex $email_list 1]]
> -    set email(name) [::json::write string $name]
>      set email(domain) [::json::write string [lindex $email_list 0]]
> -    set email_subobject [::json::write object {*}[array get email]]
>  } else {
>      set name [string trimright [lindex $detail_list 0]]
> -    set email(name) [::json::write string $name]
>      set email(domain) [::json::write string "macports.org"]
> -    set email_subobject [::json::write object {*}[array get email]]
> -}
> +set email(name) [::json::write string $name]
> +set email_subobject [::json::write object {*}[array get email]]
> ```
> 

Got it, thanks! It seems that it can be done. I will give it a try.


> My question is: does the code still work if you change `{domain:name @github}` into `{@github domain:name}`?

No, it won't. After splitting, it relies on `github username` being at `index 1` and and `email` at `index 0`.

Reply:
> > My question is: does the code still work if you change `{domain:name @github}` into `{@github domain:name}`?
> 
> No, it won't. After splitting, it relies on `github username` being at `index 1` and `email` at `index 0`.

This doesn't sound OK then. I'm not aware of any rule that dictates the github handle to come after the email address. You should iterate through entries in the list and detect github handles based on whether the first character is `@`, and for entries without `@` distinguish emails from macports usernames based on the existence of the colon. Putting email after github handle is a perfectly valid setup; it's also valid (though discouraged) to use just email or just github handle.

Reply:
Thank You for the information @mojca. I thought it was always going to be in the same format. I shall do the changes shortly.

How about this:  split at `space` and assign GitHub username if `@` is present in the entry. The other entry would then automatically be `email`, following the presence of `:` to determine the domain. And keeping in mind that entries might be missing.

Reply:
Do you need an explicit splitting at space?
Just treat entries inside maintainers the same way as `variants`, `depends_x` etc. and loop over them. Then yes, you do an if-else to check for presence of `'@'` and `':'`.